src/Extension/Traits/LocalisedCopyTrait.php

```diff
@@ -0,0 +1,132 @@
+<?php
+
+namespace TractorCow\Fluent\Extension\Traits;
+
+use SilverStripe\Versioned\Versioned;
+use TractorCow\Fluent\Extension\FluentExtension;
+use TractorCow\Fluent\Forms\CopyLocaleAction;
+
+/**
+ * Trait LocalisedCopyTrait
+ *
+ * Provides a hook which is executed when object is localised
+ * This is useful in case some action needs to be executed as a part of locale copy process
+ * examples:
+ * - related object needs to be duplicated into the new locale
+ * - event needs to be dispatched when page is localised
+ * - localised copy of a page needs to be executed outside of usual flow
+ *
+ * @see CopyLocaleAction
+ */
+trait LocalisedCopyTrait
+{
+    /**
+     * Flag indicating if the object is going through copy to locale action
+     * this is needed to determine if we need to perform localised copy
+     * this is an in-memory only field
+     *
+     * @var bool
+     */
+    private $localisedCopyActive = false;
+
+    public function getLocalisedCopyActive(): bool
+    {
+        return $this->localisedCopyActive;
+    }
+
+    public function setLocalisedCopyActive(bool $active): self
+    {
+        $this->localisedCopyActive = $active;
+
+        return $this;
+    }
+
+    /**
+     * Provides a safe way to temporarily alter the global flag
+     * useful for unit tests
+     *
+     * @param callable $callback
+     * @return mixed
+     */
+    public function withLocalisedCopyState(callable $callback)
+    {
+        $active = $this->localisedCopyActive;
+
+        try {
+            return $callback();
+        } finally {
+            $this->setLocalisedCopyActive($active);
+        }
+    }
+
+    /**
+     * Extension point in @see CopyLocaleAction::handleAction()
+     *
+     * @param string $fromLocale
+     * @param string $toLocale
+     */
+    public function onBeforeCopyLocale(string $fromLocale, string $toLocale): void
+    {
+        $this->setLocalisedCopyActive(true);
```

Comment:
Instead of a temporary variable, could we not just call executeLocalisedCopy() directly?

Reply:
The call needs to happens onBeforeWrite() because there are other cases which `onBeforeCopyLocale` doesn't cover (note that this is Localisation manager action). Cases not covered here:

* copy-to-draft button
* manual `write()` call